### PR TITLE
IGNITE-18966: Sql. Custom data types. Fix least restrictive type and nullability.

### PR DESCRIPTION
**Pull request by lowka**

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java [hunk 1]
```diff
@@ -46,7 +54,10 @@ public void dropTables() {
     public void testFilter(ColumnPair columnPair) {
         prepareTables(columnPair);
 
-        assertQuery("SELECT T11.c2 FROM T11 WHERE T11.c2 > 1.0").check();
+        String value = columnPair.lhsLiteral(0);
+        String query = format("SELECT T11.c2 FROM T11 WHERE T11.c2 > CAST({} AS {})", value, columnPair.rhs);
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java:58]: `CAST({} AS {})` -- this doesn't look like _implicit_ cast anymore

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/ImplicitCastsTest.java [hunk 2]
```diff
@@ -82,62 +82,30 @@ public void testNestedLoop(RelDataType lhs, RelDataType rhs, ExpectedTypes expec
...
+    public void testFilter(RelDataType lhs, RelDataType rhs, ExpectedTypes expected) throws Exception {
         IgniteSchema igniteSchema = new IgniteSchema("PUBLIC");
 
         addTable(igniteSchema, "A1", "COL1", lhs);
 
-        assertPlan("SELECT * FROM A1 WHERE COL1 > 1", igniteSchema, isInstanceOf(IgniteTableScan.class)
+        // Parameter types are not checked during the validation phase.
+        List<Object> params = List.of("anything");
+
+        assertPlan("SELECT * FROM A1 WHERE COL1 > CAST(? AS " + rhs + ")", igniteSchema, isInstanceOf(IgniteTableScan.class)
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/ImplicitCastsTest.java:93]: `CAST(? AS " + rhs + ")"` -- this is not _implicit_ cast

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java (hunk 1 above)
Comment (lowka) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java:58]: An implicit cast is added to the left side of an expression ( T11.c2.)

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java [hunk 3]
```diff
@@ -291,6 +291,33 @@ private RelDataType syncAttributes(
...
+            return tryCustomTypeCoercionRules(type1, type2);
+        } else if (type2 instanceof IgniteCustomType) {
+            return tryCustomTypeCoercionRules(type2, type1);
+        } else {
+            return super.commonTypeForBinaryComparison(type1, type2);
+        }
+    }
+
+    private @Nullable RelDataType tryCustomTypeCoercionRules(RelDataType type1, RelDataType type2) {
+        IgniteCustomType to = (IgniteCustomType) type1;
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java:311]: Yep. I also didn't like the way I wrote but I was not sure.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java [hunk 4]
```diff
@@ -375,28 +373,31 @@ public Type getResultClass(RelDataType type) {
...
-                    } else if (!Objects.equals(firstCustomType.getCustomTypeName(), customType.getCustomTypeName())) {
-                        // IgniteCustomType: Conversion between custom data types is not supported.
-                        return null;
+                    } else {
+                        IgniteCustomType customType = (IgniteCustomType) type;
+                        if (!Objects.equals(firstCustomType.getCustomTypeName(), customType.getCustomTypeName())) {
+                            // IgniteCustomType: Conversion between custom data types is not supported.
+                            return null;
+                        } else {
+                            return firstCustomType;
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java:388]: Thanks.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java [hunk 5]
```diff
@@ -291,6 +291,33 @@ private RelDataType syncAttributes(
...
+        // IgniteCustomType: If type1 is a custom data type that can be converted to type2,
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java:312]: Thanks. I would rewrite this comment and this method.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/ImplicitCastsTest.java (hunk 2 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/ImplicitCastsTest.java:93]: Implicit cast is added to the left side of the expression.

**Review by korlov42 — APPROVED** [review]

**Pull request merged**